vcstools/prof_utils.py

```diff
@@ -555,6 +578,470 @@ def check_clip(prof_to_check, toomuch=0.8, toolittle_frac=0., toolittle_absolute
+    _, y = sigmaClip(norm_prof, alpha=alpha)
+    noise_mean = np.nanmean(y)
+
+    # Take mean from normalised profile
+    norm_prof = norm_prof - noise_mean
+
+    # Stickel takes a numpy array in the form [[x0, y0], [x1, y1], ..., [xn, yn]]
+    xy_prof = []
+    for i, val in enumerate(norm_prof):
+        xy_prof.append([x[i], val])
+    xy_prof = np.array(xy_prof)
+
+    # Perform the stickel smooth
+    stickel = Stickel(xy_prof)
+    stickel.smooth_y(l)
+    reg_prof = stickel.yhat
+
+    # Spline, get derivative and roots
+    spline = UnivariateSpline(x, reg_prof, s=0, k=4)
+    dy_spline = spline.derivative()
+    dy_roots = dy_spline.roots()
+    dy_roots = [int(round(i)) for i in dy_roots]  # round to integer
+    spline = UnivariateSpline(x, reg_prof, s=0, k=5)
+    dy2_spline = spline.derivative().derivative()
+    dy2_roots = dy2_spline.roots()
+    dy2_roots = [int(round(i)) for i in dy2_roots]  # round to integer
+    dy2_profile = dy2_spline(x)
+
+    # Profile with spline applied
+    splined_profile = spline(x)
+    
+    # Find which dy roots resemble real signal
+    noise_trials = (2.0, 1.5, 1.0, 0.5, 0.25, 0.0)
+    for i in noise_trials: # This represents how generous we are with what is real signal
+        real_max = []
+        false_max = []
+        for root in dy_roots:
+            amp_at_root = splined_profile[root]
+            if amp_at_root < i*clip_noise:  # If the amp of the regularised profile is too small, it's not signal
+                false_max.append(root)
+            # If this is a min, we don't care about it
+            elif dy2_profile[root] > 0:
+                false_max.append(root)
+            else:  # Otherwise, this is a real max
+                real_max.append(root)
+        if i == noise_trials[-1]:
+            logger.warn("This looks like a very noisy profile. Results may be unreliable.")
+        if real_max:
+            break  # Break if anything has populated real_max
+
+
+    ##########################################################################################################
+    # We will create an over and under-estimate...
+
+    # The overestimate will mark the regions between the real max and the flanking false minima of the spline.
+    # This range should include the entire on-pulse region but also some noise
+
+    # The understimate will mark the regions between the real max and its flanking inflections.
+    # This range should include no noise but will miss some of the on-pulse
+    ##########################################################################################################
+
+    overest_on_pulse = []
+    underest_on_pulse = []
+    for root in real_max:
+        # Underestimated on-pulse
+        # Find where the index of root would be
+        dy2_roots = np.append(dy2_roots, root)
+        dy2_roots = sorted(dy2_roots)
+        i = dy2_roots.index(root)
+        dy2_roots.remove(root)
```

Comment:
I don't understand this bit. Doesn't the sorting make the index useless?

Reply:
Why would sorting make an index less functional?

Reply:
Sorting changes the index, right? Maybe you could explain why you do these couple of lines

Reply:
I am appending the first derivative roots to the second derivative roots. Then I find where the first derivative root is as an index so that I can reference the two surrounding second derivative roots. This is just a way to estimate the on-pulse.

Reply:
Alright, a strange way to do it but I see what you mean